streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -680,48 +680,106 @@ trait ZStream[-R, +E, +A] extends Serializable { self =>
-                       )
-                       .fork
-
-            // This finalizer makes sure that in all cases, the driver stops spawning new streams
-            // and the inner fibers are signalled to interrupt and actually exit.
-            _ <- ZManaged.finalizer(driver.interrupt *> interruptInners.succeed(()) *> permits.withPermits(n)(ZIO.unit))
+            _ <- self.foreachManaged { a =>
+                  for {
+                    latch <- Promise.make[Nothing, Unit]
+                    innerStream = Stream
+                      .bracket(permits.acquire)(_ => permits.release)
+                      .flatMap(_ => ZStream.fromEffect(latch.succeed(())).flatMap(_ => f(a)))
+                      .foreach(b => out.offer(Take.Value(b)).unit)
+                      .foldCauseM(
+                        cause => out.offer(Take.Fail(cause)) *> innerFailure.fail(cause),
+                        _ => ZIO.unit
+                      )
+                    _ <- (innerStream race interruptInners.await).fork
+                    // Make sure that the current inner stream has actually succeeded in acquiring
+                    // a permit before continuing. Otherwise, two bad things happen:
+                    // - we might needlessly fork inner streams without available permits
+                    // - worse, we could reach the end of the stream and acquire the permits ourselves
+                    //   before the inners had a chance to start
+                    _ <- latch.await
+                  } yield ()
+                }.foldCauseM(
+                    cause => (interruptInners.succeed(()) *> out.offer(Take.Fail(cause))).unit.toManaged_,
+                    _ =>
+                      innerFailure.await
+                      // Important to use `withPermits` here because the finalizer below may interrupt
+                      // the driver, and we want the permits to be released in that case
+                        .raceWith(permits.withPermits(n)(ZIO.unit))(
+                          // One of the inner fibers failed. It already enqueued its failure, so we
+                          // signal the inner fibers to interrupt. The finalizer below will make sure
+                          // that they actually end.
+                          leftDone =
+                            (_, permitAcquisition) => interruptInners.succeed(()) *> permitAcquisition.interrupt,
+                          // All fibers completed successfully, so we signal that we're done.
+                          rightDone = (_, failureAwait) => out.offer(Take.End) *> failureAwait.interrupt
+                        )
+                        .toManaged_
+                  )
+                  // This finalizer makes sure that in all cases, the driver stops spawning new streams
+                  // and the inner fibers are signalled to interrupt and actually exit.
+                  .ensuring(interruptInners.succeed(()) *> permits.withPermits(n)(ZIO.unit))
```

Comment:
This needs to be an `ensuringFirst`, otherwise the finalizer ordering will be wrong; `ensuring` appends a finalizer so it runs at the end. This means that the finalizer for the driver stream runs first, and only then will the inner fibers be interrupted.

You can actually keep the equivalence to the previous formulation with `ZManaged#onExitFirst`.

Reply:
Oh but the tests are still passing 🤨 Well this is interesting. I'll dig in some more.

Reply:
Ah yes. This is probably just racey. There's nothing *preventing* the inners from completing before the outer in the finalizer ordering test.

Reply:
Wait, that's why we have the `n` permits semaphore (`permits`), the driver finalizer kicks off, tells the inner fibers to terminate and waits for their termination (until it succeeds in collecting `n` permits from the semaphore), at which point all of the inner fibers will have terminated and finalized.

Or am I missing something?

Reply:
No, that's exactly it. However, consider that the driver stream has its own set of finalizers (from `self`) that need to run; for example, the stream you're flatMapping might have acquired a resource that the inner fibers are using.

When you do `ensuring`, the finalizer passed to `ensuring` is *appended* to the list of finalizers, not prepended. So the resulting set of finalizers for the driver stream is:
```
selfFinalizer *> interruptInners.succeed(()) *> permits.withPermits(n)(ZIO.unit)
```

Which means that the resources acquired in `self` would be released before the inner fibers.

Reply:
So just change `ensuring` to `ensuringFirst`? Both here and in `flatMapPar` right?

Reply:
I'd go with `.fork.onExitFirst(...)`, and interrupt the driver fiber before `interruptInners.succeed(())` (if it is Exit.Success).

Reply:
Well, I may be wrong here, but the `driver` fiber is not available in its own definition block. Doesn't `fork` on a ZManaged interrupt the fiber itself, then run all of the finalizers?

Reply:
You're right, that won't be possible with `ensuringFirst`, but `onExitFirst` lets you access the result. But I think you're right with your note about the interruption of the fiber, so `ensuringFirst` is actually sufficient. Thanks for being extra attentive here :-)